text/0000-extern-impl.md

```diff
@@ -0,0 +1,624 @@
+or traits defined in `my_definitions`, almost as if they were defined in
+`my_definitions`. If you want to use those implementations in other crates,
+however, you must also add a dependency to both `my_impls` and `my_definitions`
+(or you could re-export the definitions in `my_impls`).
+
+You may only implement for types/traits actually defined in `my_definitions`
+itself. Re-exports don't count.
+
+The dependency *must* be a path dependency to another crate defined in the same
+workspace. Cargo will not allow you specify `impl = True` for other crates.
+
+Publishing packages with `impl = True` to crates.io is not allowed.
+
+# Guide-level explanation (rustc)
+
+In current Rust, when you specify in your build system that a crate depends on
+another, this is a "use" dependency - the dependee can use definitions in the
+dependant.
+
+This RFC defines a new dependency relationship: the "impl" dependency. The
+dependee can provide implementations for types and traits defined in the
+dependant.
+
+We extend the `rustc` `--extern` flag with an additional `impl` flag: `--extern
+impl:dbtrait=dbtrait.rlib`. This means that the crate currently being compiled
+with this option, is allowed to implement traits and methods for types defined
+in the `dbtrait` crate (the *defining* crate).
+
+For example:
+```rust
+//! Defining crate
+trait DBTrait {}
+```
+```rust
+//! Implementation crate, also depends on the `mysql` crate.
+use mysql::MySql;
+
+// Allowed because we're depending on `dbtrait` with `impl`
+impl dbtrait::DBTrait for MySql {}
+```
+
+In other words, the implementing crate is considered the same as the defining
+crate as far as the orphan rule goes. Note that you may only implement traits
+and types for crates directly defined in the defining crate &mdash; you cannot
+implement re-exported types.
+
+The defining and implementing crates are not the same in other respects. They
+still have distinct crate names, defining their own crate-level namespaces. For
+example, if they both defined a type `Foo`, it would be two definitions
+`defining::Foo` and `implementing::Foo`.
+
+There are no other constraints on the defining and implementing crates - they
+can freely define new traits, types and other items, and implementations for
+those types and traits. An implementing crate can be a defining crate for
+another crate.
+
+Regardless, the definitions within the defining crate, the implementing crate,
+and between the implementing crate and defining crate must meet the current
+coherence requirements. Also if there are multiple implementation crates in
+scope in the dependency graph, they must also be coherent.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+[Coherence](https://rust-lang.github.io/chalk/book/clauses/coherence.html) is
+currently defined in terms of "local" definitions - that is, items defined with
+in the current crate.
```

Comment:
Hm, so I think it would help to clarify the terminology here a bit (I realize I am partially to blame for this since I may have advised you incorrectly on the terminology before). The linked document distinguishes between three concepts: 1. Coherence, which is a type of analysis result/a property that a dependency graph can have, 2. overlap checks, which are a particular algorithm to implement a component of coherence checks, and 3. the orphan rule checks, which are a separate algorithm that guarantee coherence when combined with overlap checks.

As a result, I would not describe coherence as being defined in terms of local definitions. The overlap checks specifically are though.

Reply:
Thanks for the clarification. I've reworded this (taking your "adoption rule" naming, since it makes sense to me now).